factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java

```diff
@@ -37,97 +43,173 @@ public class PgTransformationCache implements TransformationCache {
+  @VisibleForTesting
+  /* entry of null means read, entry of non-null means write */
+  private final Map<Key, Fact> buffer = Collections.synchronizedMap(new HashMap<>());
+
   private int maxBufferSize = 1000;
 
   @VisibleForTesting
   PgTransformationCache(
-      JdbcTemplate jdbcTemplate,
-      RegistryMetrics registryMetrics,
-      List<String> cacheKeysBatch,
-      int maxBufferSize) {
+      JdbcTemplate jdbcTemplate, RegistryMetrics registryMetrics, int maxBufferSize) {
     this.jdbcTemplate = jdbcTemplate;
     this.registryMetrics = registryMetrics;
-    this.cacheKeysBuffer = cacheKeysBatch;
     this.maxBufferSize = maxBufferSize;
   }
 
   @Override
-  public void put(@NonNull Fact fact, @NonNull String transformationChainId) {
-    String cacheKey = CacheKey.of(fact, transformationChainId);
-
-    // dup-keys can be ignored, in case another node just did the same
-
-    jdbcTemplate.update(
-        "INSERT INTO transformationcache (cache_key, header, payload) VALUES (?, ? :: JSONB, ? ::"
-            + " JSONB) ON CONFLICT(cache_key) DO NOTHING",
-        cacheKey,
-        fact.jsonHeader(),
-        fact.jsonPayload());
+  public void put(@NonNull TransformationCache.Key key, @NonNull Fact f) {
+    registerWrite(key, f);
   }
 
   @Override
-  public Optional<Fact> find(
-      @NonNull UUID eventId, int version, @NonNull String transformationChainId) {
-    String cacheKey = CacheKey.of(eventId, version, transformationChainId);
+  public Optional<Fact> find(Key key) {
+
+    Fact factFromBuffer = buffer.get(key);
+    if (factFromBuffer != null) {
+      registerAccess(key);
+      registryMetrics.count(EVENT.TRANSFORMATION_CACHE_HIT);
+      return Optional.of(factFromBuffer);
+    }
 
     List<Fact> facts =
         jdbcTemplate.query(
             "SELECT header, payload FROM transformationcache WHERE cache_key = ?",
-            new Object[] {cacheKey},
-            ((rs, rowNum) -> {
-              String header = rs.getString("header");
-              String payload = rs.getString("payload");
-
-              return Fact.of(header, payload);
-            }));
+            new Object[] {key.id()},
+            new FactRowMapper());
 
     if (facts.isEmpty()) {
       registryMetrics.count(EVENT.TRANSFORMATION_CACHE_MISS);
-
       return Optional.empty();
+    } else {
+      registerAccess(key);
+      registryMetrics.count(EVENT.TRANSFORMATION_CACHE_HIT);
+      return Optional.of(facts.get(0));
     }
+  }
+
+  @Override
+  public Set<Fact> findAll(Collection<Key> keys) {
 
-    cacheKeysBuffer.add(cacheKey);
-    if (cacheKeysBuffer.size() >= maxBufferSize) {
-      CompletableFuture.runAsync(this::flush);
+    List<Fact> facts =
+        keys.stream().map(buffer::get).filter(Objects::nonNull).collect(Collectors.toList());
+
+    if (facts.size() < keys.size()) {
+      NamedParameterJdbcTemplate named = new NamedParameterJdbcTemplate(jdbcTemplate);
+      SqlParameterSource parameters =
```

Comment:
Is there a specific reason why we (try to) fetch all transformed facts from PG instead of only the ones which are missing (from near cache)? 🤔

Reply:
test added